### ISSUE TITLE
Decide and document rules for what can be hosted as staging projects

### ISSUE THREAD

**Issue by thockin** [kubernetes/k8s.io#915]
Context: #912

**Comment by thockin** [comment]
Copying some:

The issue here is whether CNCF should be expected to foot the bill for serving other projects' container images publicly. We do this for some (coredns and etcd) because a) they are integral to k8s installs and b) they are part of CNCF.

It's not clear to me that those criteria are correct or sufficient.

We need to understand why we would say yes to one project and no to another.

Is CNCF ownership a requirement?

Is the fact that some k8s project took a dependency on it sufficient?

Should we assert that all binaries in our GCR are built by mechanisms we control and can audit?

**Comment by spiffxp** [comment]
> Is CNCF ownership a requirement?

I would argue Kubernetes ownership.  This WG's scope is about Kubernetes project infra.  If we're talking about CNCF project infra, this WG should be re-homed under CNCF.

> Is the fact that some k8s project took a dependency on it sufficient?

Not in my opinion.  I would argue Kubernetes releases are a special exception..

> Should we assert that all binaries in our GCR are built by mechanisms we control and can audit?

That would be my preference.

My strawman proposal:
- the source used to create the artifact is part of a Kubernetes subproject (owned by a SIG, or in exceptional cases by a Committee)
- the artifact is built using mechanisms wg-k8s-infra owns
- mumble something about not mirroring, with exceptions given on a case-by-case basis (e.g. Kubernetes' coredns, etcd artifacts, which are shipped as part of a Kubernetes release)

**Comment by thockin** [comment]
CoreDNS is the obvious outlier among existing staging repos.

We could fork it and cut builds from our fork.  A lot of work for low return...

**Comment by stp-ip** [comment]
On the CoreDNS side of things we have a (long term) tracking issue: #533 
TL;DR: It's more about the possibility to mirror code, and control the build (incl. process) aka have more flexibility what goes into the artifact, but not a full fork only a build wrapper repo of sorts. As we already do mirror the actual artifact and control the endpoint.

On the main issue:
From a user's point of view it seems not mirroring might lead to a bad experience with K8s. So this would mean anything that could be said is contained or needed for "plain" Kubernetes should be mirrored or at least be prevented from breaking.

One middle ground would be a container redirection idea that Tim was playing around a while back based on nginx (or a managed service should be better for our use case) while mirroring artifacts non publicly. Meaning we own the endpoint so we can always switch to publicly mirroring of some type, if anything breaks, but don't openly encourage usage of CNCF resources by providing a free and fast mirror on GCP. With the huge downside of k8s-infra now needing to maintain such a redirector.
Side effect could be a bit more insights into what various projects use as there is no specific reason other than being used within k8s to use this redirector.

**Comment by spiffxp** [comment]
I just bumped into our lack of docs on this again with CRI-O https://github.com/kubernetes/k8s.io/issues/1882

**Comment by spiffxp** [comment]
Reiterating from https://github.com/kubernetes/k8s.io/issues/915#issuecomment-702192163 what I think our policy should be

My strawman proposal:

- the source used to create the artifact is part of a Kubernetes subproject (owned by a SIG, or in exceptional cases by a Committee)
- the artifact is built using mechanisms wg-k8s-infra owns
- no mirroring, with exceptions given on a case-by-case basis to artifacts that are involved in the build, test, and release of Kubernetes (e.g. Kubernetes' coredns, etcd artifacts shipped as part of release)

I can write this up and add to whatever docs people are using to create staging repos.  What else do we need to do here?

**Comment by spiffxp** [comment]
I would be happy for this whole effort to eventually turn into something more like "wg-cncf-infra" but until that actually happens (with additional funding and staff to support the additional scope) I still think it's reasonable to explicitly disallow CNCF projects that aren't part of Kubernetes.

What do other folks think?

**Comment by dims** [comment]
@spiffxp i agree with the proposal and the caveat you outlined (better to frame it as everything originating from one of the kubernetes github orgs?)

one thing i am still not sure is the retention period. what were your thoughts there?

**Comment by thockin** [comment]
That seems like a reasonable proposal.  Does that mean CRI-O is *not*
allowed?

**Comment by spiffxp** [comment]
I put up my strawman with explicit examples in https://github.com/kubernetes/k8s.io/pull/2806

I think that would be enough to call this done

**Comment by spiffxp** [comment]
/close
Fixed by https://github.com/kubernetes/k8s.io/pull/2806
